### PR TITLE
Refactor comparison funcitons

### PR DESCRIPTION
**Pull request by Affie**
* Branch to look at comparisons in more detail, as they always seem to break

### CODE REVIEW

**Comment by Affie** [comment]
The function that broke was an easy fix, but hard to find. It is also a lot of work to maintain. So I propose using generated functions for compare overloading ==

File: src/services/CompareUtils.jl [hunk 1]
```diff
@@ -1,9 +1,40 @@
...
+TODO <:InferenceVariable: please someone confirm this, I'm not fully up to date why softtype was created.
+For now abstract `InferenceVariable`s are considered equal if they are the same type (except for labels, that I feel are deprecated)
+If your implentation has aditional properties such as `DynPose2` with `ut::Int64` (microsecond time) or support different manifolds
+implement compare if needed. softtype sounds a bit if it can be a trait, or in the future, rather a normal julia type.
+=#
+==(a::InferenceVariable,b::InferenceVariable) = typeof(a) == typeof(b)
+
+==(a::ConvolutionObject, b::ConvolutionObject) = typeof(a) == typeof(b)
+
+==(a::FunctorInferenceType, b::FunctorInferenceType) = typeof(a) == typeof(b)
```
Comment (dehann) [inline src/services/CompareUtils.jl]: just checking you know about `===`

**Comment by dehann** [comment]
sounds good, merge when ready -- doesnt seem to have any breaking changes?

**Comment by dehann** [comment]
also please add labels and milestones to this PR as appropriate :-)

**Comment by Affie** [comment]
The big thing is that you have abstract compares that fall back to just checking types. The packages that inherit then have to implement compare if they want more.

File: src/services/CompareUtils.jl (hunk 1 above)
Comment (Affie) [reply src/services/CompareUtils.jl]: Jip, but an object returned by a persistent driver will not be `===`

**Comment by Affie** [comment]
Note, this is just to develop. When merging everything to master it will unfortunately  have breaking changes that we cannot avoid. Specifically the variable softtype parameter and API standardization changes in error behavior and return types.

**Comment by dehann** [comment]
Sure, this is all for DFG v0.6, so breaking changes with deprecation warnings is fine (as much possible -- and more rigorous going forward).

**Pull request merged**